components/purchase_modal/purchase_modal.tsx

```diff
@@ -36,54 +39,65 @@ let stripePromise: Promise<Stripe | null>;
 type Props = {
     show: boolean;
     isDevMode: boolean;
-    products?: Dictionary<Product>;
+    products: Dictionary<Product> | undefined;
     contactSupportLink: string;
     contactSalesLink: string;
     isFreeTrial: boolean;
+    productId: string | undefined;
     actions: {
         closeModal: () => void;
         getCloudProducts: () => void;
         completeStripeAddPaymentMethod: (stripe: Stripe, billingDetails: BillingDetails, isDevMode: boolean) => Promise<boolean | null>;
+        subscribeCloudSubscription: (productId: string) => Promise<boolean | null>;
         getClientConfig: () => void;
         getCloudSubscription: () => void;
     };
 }
 
 type State = {
     paymentInfoIsValid: boolean;
-    productPrice: number;
     billingDetails: BillingDetails | null;
     cardInputComplete: boolean;
     processing: boolean;
+    selectedProduct: Product | null | undefined;
+}
+
+function findProductInDictionary(products: Dictionary<Product> | undefined, productId?: string): Product | null {
+    if (!products) {
+        return null;
+    }
+    const keys = Object.keys(products);
+    let selectedProduct = null;
+    if (keys.length > 1) {
+        // here find the product by the provided id, otherwise return the one with Professional in the name
+        keys.forEach((key) => {
+            if (productId && products[key].id === productId) {
+                selectedProduct = products[key];
+            } else if (!productId) {
+                selectedProduct = products[keys[0]];
+            }
```

Comment:
This check if no productId, doesn't seem related to this for each loop, should we check this at the beginning and return early?